File: src/codes/forced-compatibility/py/rfc_facade.py
class LegacyPrinter:
    def print_text(self, text: str):
        print(f"*** {text} ***")

class FooterGenerator:
    def get_footer(self) -> str:
        return "---- END ----"

class PrintFacade:
    def __init__(self):
        self.printer = LegacyPrinter()
        self.footer_gen = FooterGenerator()

    def print_with_footer(self, text: str):
        formatted = text.upper()
        self.printer.print_text(formatted)
        footer = self.footer_gen.get_footer()
        self.printer.print_text(footer)

class ReportGenerator:
    def __init__(self, facade: PrintFacade):
        self.facade = facade

    def print_summary(self, text: str):
        self.facade.print_with_footer(text)

# 利用例
facade = PrintFacade()
report = ReportGenerator(facade)
report.print_summary("月次レポート")

File: src/codes/forced-compatibility/php/rfc_facade.php
<?php

class LegacyPrinter {
  public function printText(string $text): void {
    echo "*** {$text} ***\n";
  }
}

class FooterGenerator {
  public function getFooter(): string {
    return "---- END ----";
  }
}

class PrintFacade {
  private LegacyPrinter $printer;
  private FooterGenerator $footerGen;

  public function __construct() {
    $this->printer = new LegacyPrinter();
    $this->footerGen = new FooterGenerator();
  }

  public function printWithFooter(string $text): void {
    $formatted = strtoupper($text);
    $this->printer->printText($formatted);
    $footer = $this->footerGen->getFooter();
    $this->printer->printText($footer);
  }
}

class ReportGenerator {
  public function __construct(private PrintFacade $facade) {}

  public function printSummary(string $text): void {
    $this->facade->printWithFooter($text);
  }
}

// 利用例
$facade = new PrintFacade();
$report = new ReportGenerator($facade);
$report->printSummary("月次レポート");
